SPARC: Add FPU support

martin-aberg opened the pull request:
This change adds full shared floating point support for the SPARC
architecture.

The FPU support has been tested with QEMU,
on LEON3 hardware and with the TSIM3 LEON3 simulator.

### Implementation
All SPARC floating point registers are scratch registers with respect
to function call boundaries. That means we only have to save floating
point registers when switching threads in ISR. The registers are
stored to the corresponding thread stack.

All FPU context save/restore is synchronous (not lazy).
FPU is disabled when calling ISR. Any attempt to use FPU in ISR
will generate the `fp_disabled` trap which causes Zephyr fatal error.

tbr-tt committed:
SPARC: optimized interrupt trap stores and loads

The input registers (i0..i7) are not modified by the interrupt trap
handler and are preserved by function calls. So we do not need to
store them in the interrupt stack frame.

This saves 48 bytes of stack space for each interrupted context,
and eliminates 4 double word stores and 4 double word loads.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
SPARC: optimized interrupt stack frame size

With this change we allocate stack space only for the registers we
actually store in the thread interrupt stack frame.

Furthermore, no function is called on with the interrupt context save
frame %sp so no full frame is needed here. ABI functions are called
later in the interrupt trap handler, but that is after the dedicated
interrupt stack has been installed.

This saves 96 bytes of stack space for each interrupted context.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
SPARC: add FPU support

This change adds full shared floating point support for the SPARC
architecture.

All SPARC floating point registers are scratch registers with respect
to function call boundaries. That means we only have to save floating
point registers when switching threads in ISR. The registers are
stored to the corresponding thread stack.

FPU is disabled when calling ISR. Any attempt to use FPU in ISR
will generate the fp_disabled trap which causes Zephyr fatal error.

- This commit adds no new thread state.
- All FPU contest save/restore is synchronous and lazy FPU context
  switch is not implemented.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
boards: set CPU_HAS_FPU on LEON3 soc and boards

GR716A and QEMU has FPU.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
tests/sprintf: FP tests independent of endian

The test access double variables as individual uint32_t. This
commit ensures that the correct uint32_t is accessed, based on
CONFIG_BIG_ENDIAN.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
tests/fpu_sharing: ported to SPARC

Added the bits and pieces required to run the test on SPARC.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tbr-tt committed:
tests/float_disable: ported to SPARC

Enable the test for SPARC and take into account that
arch_float_disable() returns -ENOSYS.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

tejlmand commented on tests/kernel/fpu_sharing/generic/src/test_common.h:
```diff
@@ -9,7 +9,7 @@
  * Test Thread Parameters
  */
-#define THREAD_STACK_SIZE	1024
+#define THREAD_STACK_SIZE	(1024 + CONFIG_TEST_EXTRA_STACKSIZE)
```
Seems we have both `ZTEST_STACKSIZE` and `TEST_EXTRA_STACKSIZE`.

Should this be a `(ZTEST_STACKSIZE + TEST_EXTRA_STACKSIZE)` ?
and should we consider to align those two settings ?
@galak @nashif any thoughts on this ?

galak replied:
feels like aligning the two and phasing one out if possible.

nashif replied:
ZTEST_STACKSIZE is not related here, this is for the threads that are created within the test, ZTEST_STACKSIZE is for ztest itself.

martin-aberg replied:
Thank you @tejlmand for the review. `(1024 + CONFIG_TEST_EXTRA_STACKSIZE)` is intentional.

tejlmand replied:
thanks for clarification.

ioannisg commented:
@martin-aberg this patch deserves a note in the 2.5 release notes. Adding the label, to track this. Or you can add a commit to the rel-notes already.

tbr-tt committed:
doc: release 2.5: Announce the SPARC FPU support

The SPARC architecture support added in Zephyr 2.5 provides FPU
support.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

martin-aberg commented:
Good point regarding the release notes @ioannisg. It is now added in a separate commit in this pull request

I have also rebased to the current master to make sure CI is still all right. The only change in the patch set since the initial push is the addition of the release note commit.

stephanosio commented on boards/sparc/qemu_leon3/Kconfig.board:
```diff
@@ -6,3 +6,4 @@ config BOARD_QEMU_LEON3
 	bool "QEMU LEON3 target"
 	depends on SOC_LEON3
 	select QEMU_TARGET
+	select CPU_HAS_FPU
```
@martin-aberg I am a little late to the party, but any reason you specified `CPU_HAS_FPU` at board-level instead of SoC-level as done almost everywhere else?

(came across this while reviewing an unrelated PR)
